includes/class-wc-cart.php

```diff
@@ -1120,6 +1120,20 @@ public function add_to_cart( $product_id = 0, $quantity = 1, $variation_id = 0,
 				}
 			}
 
+			// Validate variation ID.
+			if (
+				0 < $variation_id && // Only check if variation_id exists.
+				(
+					'product_variation' !== get_post_type( $variation_id ) || // Check if doesn't belong to a product variation.
+					wp_get_post_parent_id( $variation_id ) !== $product_id // Check if belongs to the selected variable product.
+				)
+			) {
+				$product = wc_get_product( $product_id );
+
+				/* translators: 1: product link, 2: product name */
+				throw new Exception( sprintf( __( 'The selected product isn\'t a variation of %2$s, please choose product options by visiting <a href="%1$s" title="%2$s">%2$s</a>.', 'woocommerce' ), esc_url( $product->get_permalink() ), esc_html( $product->get_name() ) ) );
+			}
+
```

Comment:
This looks clear and clean :star_struck: 

One thing I wondered is if we could use more of the existing logic in this method? For example, right after [lines 1035-1037](https://github.com/woocommerce/woocommerce/pull/29208/files#diff-6feb15311be1c9e4ffd7de45997d560228de3e4bd97fa057c7006074d121847bR1035-R1037):

* We already know a `$variation_id` was passed in.
* We already have the variation object contained within `$product_data`.
* We have the parent object (for the variation) within `$parent_data`.

So, we could perhaps straight up check if `$product_data->get_parent_id() !== $product_id` and, rather than creating a new _$product_ object (to use in the exception message), we could use the existing `$parent_data` object.

Definitely not essential, though—just an option :-)

Reply:
Thanks for looking into this. I agree about `$product_data`, I'll make those changes, but we may not have a `$parent_data` at this point, we we still need to load.

Reply:
Note that we'll need to keep `0 < $variation_id` since we should only check when this value is passed, and not always when some product have a parent, since it may conflict with some 3rd party plugins. After that is already save to use `$product_data`, since we have already loaded from `$variation_id` as you mentioned.